src/libproc_macro/lib.rs

```diff
@@ -211,12 +213,130 @@ impl Span {
         ::__internal::with_sess(|(_, mark)| Span(mark.expn_info().unwrap().call_site))
     }
 
+    /// The original source file into which this span points.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn source_file(&self) -> SourceFile {
+        SourceFile {
+            filemap: __internal::lookup_char_pos(self.0.lo()).file,
+        }
+    }
+
+    /// Get the starting line/column in the source file for this span.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn start(&self) -> LineColumn {
+        let loc = __internal::lookup_char_pos(self.0.lo());
+        LineColumn {
+            line: loc.line,
+            column: loc.col.to_usize()
+        }
+    }
+
+    /// Get the ending line/column in the source file for this span.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn end(&self) -> LineColumn {
+        let loc = __internal::lookup_char_pos(self.0.hi());
+        LineColumn {
+            line: loc.line,
+            column: loc.col.to_usize()
+        }
+    }
+
+    /// Create a new span encompassing `self` and `other`.
+    ///
+    /// Returns `None` if `self` and `other` are from different files.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn join(&self, other: Span) -> Option<Span> {
+        let self_loc = __internal::lookup_char_pos(self.0.lo());
+        let other_loc = __internal::lookup_char_pos(self.0.lo());
+
+        if self_loc.file.name != other_loc.file.name { return None }
+
+        Some(Span(self.0.to(other.0)))
+    }
+
     diagnostic_method!(error, Level::Error);
     diagnostic_method!(warning, Level::Warning);
     diagnostic_method!(note, Level::Note);
     diagnostic_method!(help, Level::Help);
 }
 
+/// A line-column pair representing the start or end of a `Span`.
+#[unstable(feature = "proc_macro", issue = "38356")]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct LineColumn {
+    /// The 1-indexed line in the source file on which the span starts or ends (inclusive).
+    line: usize,
+    /// The 0-indexed column (in UTF-8 characters) in the source file on which
+    /// the span starts or ends (inclusive).
+    column: usize
+}
+
+/// The source file of a given `Span`.
+#[unstable(feature = "proc_macro", issue = "38356")]
+#[derive(Clone)]
+pub struct SourceFile {
+    filemap: Rc<FileMap>,
+}
+
+impl SourceFile {
+    /// Get the path to this source file.
+    ///
+    /// ### Note
+    /// If the code span associated with this `SourceFile` was generated by an external macro, this
+    /// may not be an actual path on the filesystem. Use [`is_real`] to check.
+    ///
+    /// [`is_real`]: #method.is_real
+    # [unstable(feature = "proc_macro", issue = "38356")]
+    pub fn as_path(&self) -> &Path {
+        Path::new(&self.filemap.name)
```

Comment:
Will this path be used to open files? Note that `filemap.name` is not a real path when `filemap.name_was_remapped` is true (using `-Z remap-path-prefix-*`). It will look like a real path though, so if it's just used for display purposes then that's fine.

Reply:
So `is_real()` should probably return `filemap.is_real_file() && filemap.name_was_remapped`?

Reply:
If it needs to be a real path that can be opened, then maybe you need the `filemap.path` field that I'm adding in #44940. If users of this API need to be able to open the file, then they need to be able to open the file even when using `-Z remap-path-prefix-*`.

Reply:
The main use case is to refer to files in error messages, so they should make sense to the user.

Reply:
Ok, in that case `filemap.name` is correct, and you don't need to check `name_was_remapped`. Maybe it should be returning a `str` instead of a `Path`, so that users are less likely to wrongly open it.

Reply:
Yeah I can't really dispute that.

Reply:
I'm very interested in opening files through the path I get from the span. Why shouldn't we allow it?

Reply:
It's not that it's not allowed, it's that the path may not be valid on the filesystem. If it was remapped it may not point to an actual file, and if it was generated by a macro in an external crate then the source file will be `<[crate name] macros>`.

Reply:
Ah I see. But I don't know what this "remapping" is and when it occurs. Could you point me to a resource where I could learn about it? Or maybe you could explain it in the doc comment for `is_real()`? It would fit there, right? :)

Reply:
I assume it's used by some build systems. I do mention it in the `as_str()` docs, think it needs to be duplicated on `is_real()`?

Reply:
The remapping is for when the source package is going installed at a location that is different from where it is built. eg I may compile in /home/philipc, but the source package will be installed in /usr/src. It's also useful for reproducible builds. See #38322 for the feature request.

What it means is that the paths in the output binaries should use the remapped location (such as in debuginfo or panic messages). We're also using the remapped path for error messages during compilation, but I'm less sure about why that is useful (maybe it's just for anonymization).

But if you need to open a file during compilation then you need to use the path that isn't remapped, so maybe we need to add an API for that too. Otherwise we'll get into a situation where crates will fail to compile when remapping is used.

Reply:
Thanks for explaining! I guess we really should add an API to retrieve the original path in long term. The remap-flag should never break code that compiled before, I guess. But maybe this can be done later, as this PR has seen already quite a bit of iteration?

Reply:
@philipc something just occurred to me: the way it's currently implemented, remapping seems to be an all-or-nothing thing, so can it just be a global flag instead of per `Filemap`?

Reply:
It's not all-or-nothing: it's possible the remap prefix will only match some files, or there may be multiple prefixes that match different files. I don't know if this will happen in practice, but the compiler currently supports it.